### PR TITLE
Proposal: Add `filament:optimize` and `filament:optimize-clear` commands

### PR DESCRIPTION
**Pull request by dissto**
## Description

This PR adds two new commands under the filament namespace: 
- `php artisan filament:optimize`
- `php artisan filament:optimize-clear`

Its just a wrapper for the 2 recommended commands `icons:cache` and `filament:cache-components`

## Visual changes

https://github.com/user-attachments/assets/624a1015-629c-4d7b-a084-12bd83ea8470

## Functional changes

- [x] Code style has been fixed by running the `composer cs` command.
- [x] Changes have been tested to not break existing functionality.
- [x] Documentation is up-to-date.

### CODE REVIEW

File: packages/support/src/Commands/OptimizeCommand.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+use Illuminate\Console\Command;
+use Symfony\Component\Console\Attribute\AsCommand;
+
+#[AsCommand(name: 'filament:optimize')]
+class OptimizeCommand extends Command
+{
+    protected $signature = 'filament:optimize';
+
+    protected $description = 'Cache components and blade icons to increase performance.';
```
Comment (danharrin) [inline packages/support/src/Commands/OptimizeCommand.php]:
```suggestion
    protected $description = 'Cache components and Blade icons to improve performance.';
```

File: packages/support/src/Commands/OptimizeCommand.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    public function handle(): int
+    {
+        $this->components->info('Caching components and blade icons.');
```
Comment (danharrin) [inline packages/support/src/Commands/OptimizeCommand.php]:
```suggestion
        $this->components->info('Caching components and Blade icons...');
```

File: packages/support/src/Commands/OptimizeCommand.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+        collect([
+            'Caching components' => fn () => $this->callSilent('filament:cache-components') == 0,
+            'Caching blade icons' => fn () => $this->callSilent('icons:cache') == 0,
```
Comment (danharrin) [inline packages/support/src/Commands/OptimizeCommand.php]:
```suggestion
            'Caching Blade icons' => fn (): bool => $this->callSilent('icons:cache') === static::SUCCESS,
```

File: packages/support/src/Commands/OptimizeCommand.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    protected $signature = 'filament:optimize';
+
+    protected $description = 'Cache components and blade icons to increase performance.';
+
+    public function handle(): int
+    {
+        $this->components->info('Caching components and blade icons.');
+
+        collect([
+            'Caching components' => fn () => $this->callSilent('filament:cache-components') == 0,
```
Comment (danharrin) [inline packages/support/src/Commands/OptimizeCommand.php]:
```suggestion
            'Caching components' => fn (): bool => $this->callSilent('filament:cache-components') === static::SUCCESS,
```

File: packages/support/src/Commands/OptimizeClearCommand.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+use Illuminate\Console\Command;
+use Symfony\Component\Console\Attribute\AsCommand;
+
+#[AsCommand(name: 'filament:optimize-clear')]
+class OptimizeClearCommand extends Command
+{
+    protected $signature = 'filament:optimize-clear';
+
+    protected $description = 'Remove cached components and blade icons.';
```
Comment (danharrin) [inline packages/support/src/Commands/OptimizeClearCommand.php]:
```suggestion
    protected $description = 'Remove cached components and Blade icons.';
```

File: packages/support/src/Commands/OptimizeClearCommand.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    public function handle(): int
+    {
+        $this->components->info('Clearing cached components and blade icons.');
```
Comment (danharrin) [inline packages/support/src/Commands/OptimizeClearCommand.php]:
```suggestion
        $this->components->info('Clearing cached components and Blade icons...');
```

File: packages/support/src/Commands/OptimizeClearCommand.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+        collect([
+            'Clearing cached Components' => fn () => $this->callSilent('filament:clear-cached-components') == 0,
```
Comment (danharrin) [inline packages/support/src/Commands/OptimizeClearCommand.php]:
```suggestion
            'Clearing cached components' => fn (): bool => $this->callSilent('filament:clear-cached-components') === static::SUCCESS,
```

File: packages/support/src/Commands/OptimizeClearCommand.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+            'Clearing cached Blade Icons' => fn () => $this->callSilent('icons:clear') == 0,
```
Comment (danharrin) [inline packages/support/src/Commands/OptimizeClearCommand.php]:
```suggestion
            'Clearing cached Blade icons' => fn (): bool => $this->callSilent('icons:clear') === static::SUCCESS,
```

**Review by danharrin — CHANGES_REQUESTED** [review]
Please add a check around the commands that cache components / clear cached components, so if the panel builder is not installed it only handles the icons. Otherwise the command will fail if the user runs this while not having the panel builder installed, since it is just part of the support package.

**Comment by dissto** [comment]
> Please add a check around the commands that cache components / clear cached components, so if the panel builder is not installed it only handles the icons. Otherwise the command will fail if the user runs this while not having the panel builder installed, since it is just part of the support package.

I have borrowed the [check from the install command](https://github.com/filamentphp/filament/blob/3.x/packages/support/src/Commands/InstallCommand.php#L70). Or did you have something different in mind? 

I additionally made that check into a Trait and applied all your suggestions.🤓

File: packages/support/src/Concerns/CanCacheComponents.php [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace Filament\Support\Concerns;
+
+use Filament\PanelProvider;
+
+trait CanCacheComponents
+{
+    protected function canCacheComponents(): bool
+    {
+        return class_exists(PanelProvider::class);
```
Comment (danharrin) [inline packages/support/src/Concerns/CanCacheComponents.php]: Please instead check to see if the cache-components command class exists

File: packages/support/src/Concerns/CanCacheComponents.php
Comment (danharrin) [inline packages/support/src/Concerns/CanCacheComponents.php:1]: Please move this into `Filament\Support\Commands\Concerns`

**Review by danharrin — CHANGES_REQUESTED** [review]

File: packages/support/src/Concerns/CanCacheComponents.php
Comment (danharrin) [reply packages/support/src/Concerns/CanCacheComponents.php:1]: Please rename the trait and method to `CanCachePanelComponents`

**Review by danharrin — CHANGES_REQUESTED** [review]

**Comment by dissto** [comment]
Applied the latest suggestions 🤓

**Review by danharrin — CHANGES_REQUESTED** [review]
Please swap it back so that the components are cached/cleared first before the icons so its consistent with the messages and docs

**Comment by dissto** [comment]
> Please swap it back so that the components are cached/cleared first before the icons so its consistent with the messages and docs

Got it.

**Comment by danharrin** [comment]
Thanks

**Pull request merged**